docs/proposals/09-test-framework.md

```diff
@@ -0,0 +1,129 @@
+The current testframework already contains a lot of general test functions that ease the work for writing new tests.
+However there are multiple disadvantages with the current structure of the tests and the testframework:
+1. Every new test is an own testsuite and therefore needs its own `TestDef` (https://github.com/gardener/gardener/tree/master/.test-defs). With this approach there will be hundreds of test definitions, growing with every new test (or at least new test suite).
+   But in most cases new tests do not need their own special `TestDef`: it's just the wrong scope for the testmachinery and will result in unnecessary complex testruns and configurations. In addition it would result in additional maintenance for a huge number of `TestDefs`.
+2. The testsuites currently have their own specific interface/configuration that they need in order to be executed correctly see ([Worker test](https://github.com/gardener/gardener/blob/master/.test-defs/WorkerTest.yaml#L14) or [K8s Update test](https://github.com/gardener/gardener/blob/master/.test-defs/ShootKubernetesUpdateTest.yaml#L14)).
+  Consequently the configuration has to be defined in the testruns which result in one step per test with their very own configuration which means that the testmachinery cannot simply select testdefinitions by label.
+  As the testmachinery cannot make use of its ability to run labeled tests (e.g. run all tests labeled `default`), the testflow size increases with every new tests and the testruns have to be manually adjusted with every new test.
+3. The current gardener test framework contains multiple test operations where some are just used for specific tests (e.g. `plant_operations`) and some are more general (`garden_operation`). Also the functions offered by the operations vary in their specialization as some are really specific to just one test e.g. shoot test operation with `WaitUntilGuestbookAppIsAvailable` whereas others are more general like `WaitUntilPodIsRunning`.<br>
+   This structure makes it hard for developers to find commonly used functions and also hard to integrate as the common framework grows with specialized functions.
+
+## Goals
+In order to clean the testframework, make it easier for new developers to write tests and easier to add and maintain test execution within the testmachinery, the following goals are defined:
+- Have a small number of test suites (gardener, shoots see [test flavors](#test_flavors)) to only maintain a fixed number of testdefinitions.
+- Use ginkgo test labels (inspired by the k8s e2e tests) to differentiate test behavior, test execution and test importance.
+- Use standardized configuration for all tests (differ depending on the test suite) but provide better tooling to dynamically read additional configuration from configuration files like the `cloudprofile`.
+- Clean the testframework to only contain general functionality and keep specific functions inside the tests
```

Comment:
I would add as a goal to keep the test execution simple and stupid. Sometimes I locally run some integration tests to validate PRs, and currently create shoot test for example requires 15 flags to be specified. I liked the approach with the yaml file that you could pass directly the Shoot spec to be applied. WDYT?

Reply:
great idea. I would suggest something similar to [viper](https://github.com/spf13/viper) so that configuration can be provided via commandline and file.